espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ComposeRequestBuilder.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.helpers
+
+import androidx.compose.ui.test.SemanticsMatcher
+import androidx.compose.ui.test.hasAnyAncestor
+import androidx.compose.ui.test.hasContentDescription
+import androidx.compose.ui.test.hasTestTag
+import androidx.compose.ui.test.hasText
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidSelectorException
+import io.appium.espressoserver.lib.handlers.exceptions.StaleElementException
+import io.appium.espressoserver.lib.model.Locator
+import io.appium.espressoserver.lib.model.Strategy
+
+object ComposeRequestBuilder {
+
+    private fun leafLocatorToMatcher(locator: Locator): SemanticsMatcher =
+        when (locator.using) {
+            Strategy.VIEW_TAG -> hasTestTag(locator.value!!)
+            Strategy.TEXT -> hasText(locator.value!!)
+            Strategy.LINK_TEXT -> hasText(locator.value!!)
+            Strategy.ACCESSIBILITY_ID -> hasContentDescription(locator.value!!)
+            else -> throw InvalidSelectorException(
+                "Can't use non-Compose selectors. " +
+                        "only ${Strategy.VIEW_TAG}, ${Strategy.TEXT}, ${Strategy.LINK_TEXT} and " +
+                        "${Strategy.ACCESSIBILITY_ID} are supported for now"
+            )
+        }
+
+    fun locatorToMatcher(locator: Locator): SemanticsMatcher {
+        val parent = locator.elementId
+        val parentMatcher: SemanticsMatcher? =
+            parent?.let {
+                val parentLocator = ComposeViewCache.get(parent)
+                    ?: throw StaleElementException("Parent locator $parent is not found in cache")
```

Comment:
has either expired or did not exist in the cache

Reply:
changed